### PR TITLE
Add notify implementation

### PR DESCRIPTION
**Pull request by Smephite**
Add script to implement weekly notification with a XDR to kalepail for reward distribution.

This implementation will _not_ pose a bug due to multiple instances of a discord bot (web api ftw)

### CODE REVIEW

File: stellar_helpers.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
+from decimal import Decimal
+from stellar_sdk import TransactionBuilder, Server, Network, Keypair, Account, asset
+import os
+import sys
+from stellar_sdk.operation.create_claimable_balance import Claimant
+from stellar_sdk.time_bounds import TimeBounds
+from stellar_sdk.xdr.base import String
+
+STELLAR_USE_TESTNET = 'USE_STELLAR_TEST_NET' in os.environ
+STELLAR_ENDPOINT = "https://horizon.stellar.org"
```
Comment (MatejMecka) [inline stellar_helpers.py:10]:
```suggestion
STELLAR_ENDPOINT = "https://horizon-testnet.stellar.org" if STELLAR_USE_TESTNET else "https://horizon.stellar.org"
```

File: stellar_helpers.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+STELLAR_PASSPHRASE = Network.PUBLIC_NETWORK_PASSPHRASE
```
Comment (MatejMecka) [inline stellar_helpers.py:11]:
```suggestion
STELLAR_PASSPHRASE = Network.TESTNET_NETWORK_PASSPHRASE if STELLAR_USE_TESTNET else Network.PUBLIC_NETWORK_PASSPHRASE
```

File: stellar_helpers.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+if STELLAR_USE_TESTNET:
```
Comment (MatejMecka) [inline stellar_helpers.py:13]: This can all be replaced in one line :D left you some suggestions

File: stellar_helpers.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        reward = round(rewarded[1], 7)
+        
+        if rewarded >= 0:
+            continue
+
+        tx.append_create_claimable_balance_op(
+            asset=asset.Asset.native(), 
+            amount=str(reward),
+            claimants=[Claimant(rewarded[0])]
+            )
```
Comment (MatejMecka) [inline stellar_helpers.py:61]:
```suggestion
         )
```

File: stellar_helpers.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    takes tewardee (array of tuples (pub_key, amount)) where pub_key is str and amount is float
+    returns XDR as string
+    """
+    source_acc = Account(account_id=PUBLIC_KEY, sequence=1) # fetch sequence or will it be used with a fee bump?
+
+    tx = TransactionBuilder(
+        source_account=source_acc,
+        network_passphrase=STELLAR_PASSPHRASE,
+        base_fee=100)\
+            .add_text_memo("Laumenaut reward!")
```
Comment (MatejMecka) [inline stellar_helpers.py:49]:
```suggestion
            .add_text_memo("Lumenaut reward!")
```

File: helper.py [hunk 6]
```diff
@@ -1,3 +1,30 @@
+from datetime import datetime, timedelta
+
+def setup_db(conn):
+    c = conn.cursor()
+    c.execute('''CREATE TABLE IF NOT EXISTS votes_history (id INTEGER AUTO_INCREMENT PRIMARY KEY, user_id INTEGER, message_id INTEGER, backer INTEGER, vote_time DATETIME DEFAULT CURRENT_TIMESTAMP)''')
+    c.execute('''CREATE TABLE IF NOT EXISTS users(user_id INTEGER PRIMARY KEY, stellar_account VARCHAR(55))''')
+def getUserPubKey(conn, user):
```
Comment (MatejMecka) [inline helper.py:7]: New line :3

**Review by MatejMecka — CHANGES_REQUESTED** [review]
Couple of Changes, nothing major :3

File: stellar_helpers.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    Generates transaction XDR full of claimable balance operations
+    takes tewardee (array of tuples (pub_key, amount)) where pub_key is str and amount is float
+    returns XDR as string
+    """
+    source_acc = Account(account_id=PUBLIC_KEY, sequence=1) # fetch sequence or will it be used with a fee bump?
+
+    tx = TransactionBuilder(
+        source_account=source_acc,
+        network_passphrase=STELLAR_PASSPHRASE,
+        base_fee=100)\
```
Comment (Smephite) [inline stellar_helpers.py:44]: fetch base fee here or will this be bumped anyway? (same with sq above)

File: handle_distribution.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    if len(payoutUser) == 0:
+        return "No eligable users this week!"
+
+    payouts = []
+
+    for user in payoutUser:
+        payout = user[1] / sumVotes * pricepot
+        payouts.append((user[2], payout))
+        
+    tx_xdr = generate_reward_tx(payouts)
```
Comment (Kanaye) [inline handle_distribution.py:51]: Not exactly this line. And probably not important for now. But well we'll probably crash, if/when we hit more eligable users than the op-limit per tx. Probably the best thing to do right now: Check for that and send a message "Lumenauts you F*** up"

File: stellar_helpers.py (hunk 7 above)
Comment (Kanaye) [reply stellar_helpers.py:48]: We should probably just fetch the base fee tbh. While it's not a huge trouble to wrap it in a fee bump, it's an extra step we can leave out.

Comment (Smephite) [reply stellar_helpers.py:48]: The way it is implemented atm a fee bump is required as the seq number is set to 1

Comment (Kanaye) [reply stellar_helpers.py:48]: Oh didn't notice that. Not that a fee bump wouldn't help us in that case, as it just bumps the fee. The sequence number is still consumed from the source account of the inner tx

Comment (Smephite) [reply stellar_helpers.py:48]: In that case just assume the seq won't change inbetween issuing this tx and Tyler transmitting it?

Comment (Kanaye) [reply stellar_helpers.py:48]: Well be using a dedicates account for this anyways. So I don't think the suquence number will change in between.
But we could again just ask Tyler 😉

File: discord_helpers.py [hunk 9]
```diff
@@ -22,3 +27,68 @@ def hasRole(roles, REQUIRED_ROLE_ID):
...
+    last_tx_date = fetch_last_tx()
+
+    if last_tx_date == None:
+        return "Failed to load past transactions of reward account! Does it exist?"
+
+    leaderboard_rows = fetchLeaderboard(conn, last_tx_date, datetime.now())
+    user_rows = fetchUserPubKeys(conn)
+    sumVotes = 0
+
+    BASE_FEE = 10000
```
Comment (MatejMecka) [inline discord_helpers.py:41]: Maybe this can be an env variable or at the top?

File: discord_helpers.py [hunk 10]
```diff
@@ -22,3 +27,68 @@ def hasRole(roles, REQUIRED_ROLE_ID):
...
+    if len(payoutUser) > 100:
+        return "Wow! There are a lot of eligible lumenauts (>100). We should upgrade our code to handle this case..."
+    payouts = []
+
+    pricepot -= len(payoutUser) # base reserve * 2 (2 claimaints for each claimableBalance)
+
+    if pricepot <= 0:
+        return f"Balance of Pricepot is not high enough to support {len(payoutUser)} eligible lumenauts!"
+
+
```
Comment (MatejMecka) [inline discord_helpers.py:74]: One new line

File: discord_helpers.py [hunk 11]
```diff
@@ -22,3 +27,68 @@ def hasRole(roles, REQUIRED_ROLE_ID):
...
+    for user in payoutUser:
+        payout = user[1] / sumVotes * pricepot
+        payouts.append((user[2], payout))
+        
+    tx_xdr = generate_reward_tx(payouts, BASE_FEE)
+
+    if tx_xdr == None: 
+        return f"Failed to load reward account!"
+
+
```
Comment (MatejMecka) [inline discord_helpers.py:84]: 1 new line

**Review by MatejMecka — CHANGES_REQUESTED** [review]
Almost done :D

File: discord_helpers.py (hunk 9 above)
Comment (Smephite) [reply discord_helpers.py:41]: +1

**Comment by Smephite** [comment]
- [ ] Fix bug

Any transaction will be used to determine last TX date, thus payment transaction to stock up account will overwrite date -> votes from last distribution until stocking up are lost

**Review by MatejMecka — APPROVED** [review]
LGTM! Thanks for implementing all this! <3

**Pull request merged**